### PR TITLE
Support passing string type numbers, geographic coordinates and datetimes

### LINKED ISSUE #633: Pass string-type x and y to text or other plotting functions
**Description of the problem**

PR #480 mention that, *x* and *y* of the `text()` function can accept int, float or str. It's true for PyGMT v0.1.2, but since PyGMT v0.2.0, it's no longer possible to pass string-type *x* or *y*.

I haven't looked into the codes to check why it works for v0.1.2. The question is, do we want to accept string-type *x* and *y* coordinates?

**Full code that generated the error**

```python
import pygmt

fig = pygmt.Figure()

fig.basemap(region=[0, 10, 0, 10], frame=True, projection='X10c')
fig.text(x='5.0', y=5, text='TEXT')
fig.show()
```

**Full error message**

```
Traceback (most recent call last):
  File "/Users/seisman/Gits/gmt/pygmt/pygmt/clib/session.py", line 727, in _check_dtype_and_dim
    array = np.asarray(array, dtype=np.datetime64)
  File "/Users/seisman/.anaconda/lib/python3.8/site-packages/numpy/core/_asarray.py", line 85, in asarray
    return array(a, dtype, copy=False, order=order)
ValueError: Error parsing datetime string "5.0" at position 1

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "test.py", line 6, in <module>
    fig.text(x='5.0', y=5, text='TEXT')
  File "/Users/seisman/Gits/gmt/pygmt/pygmt/helpers/decorators.py", line 270, in new_module
    return module_func(*args, **kwargs)
  File "/Users/seisman/Gits/gmt/pygmt/pygmt/helpers/decorators.py", line 407, in new_module
    return module_func(*args, **kwargs)
  File "/Users/seisman/Gits/gmt/pygmt/pygmt/base_plotting.py", line 1213, in text
    with file_context as fname:
  File "/Users/seisman/.anaconda/lib/python3.8/contextlib.py", line 113, in __enter__
    return next(self.gen)
  File "/Users/seisman/Gits/gmt/pygmt/pygmt/clib/session.py", line 1173, in virtualfile_from_vectors
    self.put_vector(dataset, column=col, vector=array)
  File "/Users/seisman/Gits/gmt/pygmt/pygmt/clib/session.py", line 777, in put_vector
    gmt_type = self._check_dtype_and_dim(vector, ndim=1)
  File "/Users/seisman/Gits/gmt/pygmt/pygmt/clib/session.py", line 729, in _check_dtype_and_dim
    raise GMTInvalidInput(
pygmt.exceptions.GMTInvalidInput: Unsupported numpy data type '<class 'numpy.str_'>'.
```

**System information**

Please paste the output of `python -c "import pygmt; pygmt.show_versions()"`:

```
PyGMT information:
  version: v0.2.0+16.g7bedf487
System information:
  python: 3.8.3 (default, Jul  2 2020, 11:26:31)  [Clang 10.0.0 ]
  executable: /Users/seisman/.anaconda/bin/python
  machine: macOS-10.15.6-x86_64-i386-64bit
Dependency information:
  numpy: 1.18.5
  pandas: 1.0.5
  xarray: 0.16.1
  netCDF4: 1.5.3
  packaging: 20.4
  ghostscript: 9.53.1
  gmt: 6.2.0_3f85983_2020.09.26
GMT library information:
  binary dir: /Users/seisman/.anaconda/bin
  cores: 8
  grid layout: rows
  library path: /Users/seisman/local/GMT/lib/libgmt.dylib
  padding: 2
  plugin dir: /Users/seisman/local/GMT/lib/gmt/plugins
  share dir: /Users/seisman/local/GMT/share
  version: 6.2.0
```

### LINKED ISSUE #647: Support passing in degrees, minutes, seconds (DMS) geographical coordinates
**Description of the desired feature**

In GMT, geographical coordinates can be given in floating points or many [different geographical formats](https://docs.generic-mapping-tools.org/latest/cookbook/features.html#input-data-formats) (e.g., `23:30:30E`), but PyGMT doesn't support these common geographical formats yet.

Here is a script to show what it looks like:

```python
import pygmt

fig = pygmt.Figure()
fig.basemap(region=[-10, 10, -10, 10], projection='M10c', frame=True)
fig.plot(x=5.0, y=5.0, style='c0.5c', color='red')
fig.plot(x='3:30W', y='5:30W', style='c0.5c', color='blue')
fig.show()
```

**Expected output:**

image

**Actual output**

```
Traceback (most recent call last):
  File "/Users/seisman/Gits/gmt/pygmt/pygmt/clib/session.py", line 728, in _check_dtype_and_dim
    array = np.asarray(array, dtype=np.datetime64)
  File "/Users/seisman/.anaconda/lib/python3.8/site-packages/numpy/core/_asarray.py", line 85, in asarray
    return array(a, dtype, copy=False, order=order)
ValueError: Error parsing datetime string "3:30W" at position 1

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "test.py", line 6, in <module>
    fig.plot(x='3:30W', y='5:30W', style='c0.5c', color='blue')
  File "/Users/seisman/Gits/gmt/pygmt/pygmt/helpers/decorators.py", line 270, in new_module
    return module_func(*args, **kwargs)
  File "/Users/seisman/Gits/gmt/pygmt/pygmt/helpers/decorators.py", line 411, in new_module
    return module_func(*args, **kwargs)
  File "/Users/seisman/Gits/gmt/pygmt/pygmt/base_plotting.py", line 720, in plot
    with file_context as fname:
  File "/Users/seisman/.anaconda/lib/python3.8/contextlib.py", line 113, in __enter__
    return next(self.gen)
  File "/Users/seisman/Gits/gmt/pygmt/pygmt/clib/session.py", line 1174, in virtualfile_from_vectors
    self.put_vector(dataset, column=col, vector=array)
  File "/Users/seisman/Gits/gmt/pygmt/pygmt/clib/session.py", line 778, in put_vector
    gmt_type = self._check_dtype_and_dim(vector, ndim=1)
  File "/Users/seisman/Gits/gmt/pygmt/pygmt/clib/session.py", line 730, in _check_dtype_and_dim
    raise GMTInvalidInput(
pygmt.exceptions.GMTInvalidInput: Unsupported numpy data type '<class 'numpy.str_'>'.
```

**Possible solutions**

Here are two possible solutions in my mind:

1. Currently, `GMT_Put_Vector` doesn't support string-like geographical coordinates. It could be a missing feature of the GMT API. With the new data type added, PyGMT still needs to determine if a string is a valid geographical coordinate before passing to GMT.
2. PyGMT may provide a utility function to convert these geographical formats into floating points data. In this case, we don't need the feature in the GMT API. However, if I understand it correctly, geographical coordinates like `3:30W` is still a little different from floating points `-3.5`. For `3:30W`, GMT knows the input is geographical coordinates and may have more automatic settings for geographical coordinates.

**Are you willing to help implement and maintain this feature?** Yes, but contributions are welcome.

### PR DESCRIPTION
**Pull request by seisman**
**Description of proposed changes**

This PR adds support for passing string type data to GMT via `GMT_Put_Vectors`. Supported string type data include:

- numeric strings (e.g., `"0.5"`)
- geographic coordinates (e.g,  `"3:30W"`, `"3:30N"`)
- datetime strings (e.g., `2010-01-01T00:00`)

This PR should address the bug/feature request in #647 and #633.

To test this branch, you need GMT dev versions (i.e., GMT>=6.2.0). You can install the latest GMT dev versions using:
```
conda install -c conda-forge/label/dev gmt
```

Example script for testing:
```python
import pygmt

fig = pygmt.Figure()
fig.basemap(region=[-10, 10, -10, 10], projection='M10c', frame=["WSEN", "afg1"])
fig.plot(x=5.0, y=5.0, style='c0.2c', color='red')
fig.plot(x='-5.0', y='-5.0', style='c0.2c', color='blue')
fig.plot(x='5:30W', y='5:30N', style='c0.2c', color='yellow')
fig.savefig("map.png")
```

Expected and actual output:
image

Fixes #647.
Fixes #633.

**Reminders**

- [ ] Run `make format` and `make check` to make sure the code follows the style guide.
- [ ] Add tests for new features or tests that would have caught the bug that you're fixing.
- [ ] Add new public functions/methods/classes to `doc/api/index.rst`.
- [ ] Write detailed docstrings for all functions/methods.
- [ ] If adding new functionality, add an example to docstrings or tutorials.

**Slash Commands**

You can write slash commands (`/command`) in the first line of a comment to perform
specific operations. Supported slash commands are:

- `/format`: automatically format and lint the code
- `/test-gmt-dev`: run full tests on the latest GMT development version

### CODE REVIEW

**Comment by weiji14** [comment]
Can you add a doc API line at https://github.com/GenericMappingTools/pygmt/blame/v0.3.0/doc/api/index.rst#L214 for `clib.Session.put_strings`? Just realized we forgot to add it during #520.

**Comment by seisman** [comment]
> Can you add a doc API line at [`v0.3.0`/doc/api/index.rst#L214 (blame)](https://github.com/GenericMappingTools/pygmt/blame/v0.3.0/doc/api/index.rst#L214) for `clib.Session.put_strings`? Just realized we forgot to add it during #520.

Done in 24ad0afd69e0d958e4a9c469b48354178b18debb

**Comment by seisman** [comment]
See https://github.com/GenericMappingTools/pygmt/actions/runs/609196993 for the tests with GMT dev versions.

All the tests (including the new one) pass. 

Tests with GMT 6.1.1 fails, because in the old codes, datetime strings can processed by `pd.to_datetime`, but now datetime strings are passed to GMT directly. As expected, passing datetime strings to GMT works for GMT dev but not GMT 6.1.1, so we can't merge the PR unless GMT 6.2.0 is released.

File: pygmt/tests/test_clib_put_vector.py [hunk 1]
```diff
@@ -90,6 +90,62 @@ def test_put_vector_mixed_dtypes():
...
+        [-10, 30.5, -30.5, 40.508472],  # longitudes
+        [10, -30.50, 30.5, -40.508472],  # latitudes
+        # datetimes
+        ["2021-02-03", "2021-02-03T04", "2021-02-03T04:05:06.700", "T04:50:06.700"],
+    ]
+
+    expected_dtypes = [np.double, np.double, np.double, np.str_]
+
+    for i, j in itertools.combinations_with_replacement(range(3), r=2):
+        # TODO: Change range(3) to range(4)
```
Comment (seisman) [inline pygmt/tests/test_clib_put_vector.py:118]: The idea of the test is to pass different combinations of strings to GMT, let GMT write it to a temporary file, and check if the output is the same as expected (i.e., if `vector=expected_vetcor`.

Still has trouble testing datetime strings, when the text file contains strings,
```
dtype=[("x", np.str_), ("y", np.str_)],
```
doesn't works as I expect.

Comment (seisman) [reply pygmt/tests/test_clib_put_vector.py]: Fixed in bf7587cb2ca1b34ade13f0239fc078ca4279c332, but could still be improved.

**Review by weiji14 — APPROVED** [review]
Great to finally see this come through!

**Pull request merged**